doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,154 @@
+The user must have a `READ` privilege on all objects.
+
+Definitions:
+- **physical read frontier**: the read frontier of an object excluding `HOLD`s and any running queries or transactions.
+- **logical read frontier**: the read frontier of an object including `HOLD`s and any running queries or transactions.
+
+If `AT <time>` is specified, the read hold is created at that time.
+It is an error if that time is less than any object's logical frontier.
+Due to the use of the logical frontier, a new `HOLD` can be created while at the time of an existing `HOLD` (say, because the object set of a `HOLD` needs to be changed by creating a new one).
+If `AT <time>` is not specified, Materialize automatically chooses the least common physical read frontier of all objects.
+
+`HOLD`s are destroyed by:
+
+```
+DROP HOLD <name>
+```
+
+A `DROP` will release its current read hold.
+In-progress `SUBSCRIBE`s and transactions will not be affected after a `DROP`, but will continue to run as expected.
+Those will still be aquiring their own read holds in addition to the `HOLD` object.
+
+Creating a `HOLD` will increase resource consumption (mostly memory) in proportion to the amount of time it is held for.
+Users must periodically advance the `HOLD` to a more recent time to allow for resources to be released.
+
+`HOLD`s are advanced by:
+
+```
+ALTER HOLD <name> ADVANCE [TO <time>]
+```
+
+If `TO <time>` is specified, the new time will be created at that time.
+It is an error if that time is less than any object's logical read frontier.
+If `TO <time>` is not specified, the new time is the least common physical read frontier of all objects.
+The `HOLD` releases its current read hold and aquires a new read hold at the new time.
+
+The `ERROR_IF_OLDER_THAN` option (required, but defaults to `'24h'`) is a duration.
+If difference between the least common physical read frontier of the objects and the `HOLD` time is greater than `ERROR_IF_OLDER_THAN`, the `HOLD`'s read hold is released and the `HOLD` enters an error state.
+(Transactions and `SUBSCRIBE` are not affected, as in the `ALTER HOLD ADVANCE` case.)
+In the error state, `ALTER HOLD ADVANCE` will return an error.
+`SELECT` and `SUBSCRIBE` `AS OF` a time that is between the erroring `HOLD`'s time and the least common physical read frontier will return a helpful error message that the `HOLD` errored because it was not advanced.
+
+It is an error to `DROP` an object that is part of a `HOLD` unless the `DROP` is `CASCADE`.
+If the `DROP` is `CASCADE`, any `HOLD`s on the dropped object are also dropped (even if the `HOLD` is also over other objects).
+
+### Monitoring
+
+To monitor `HOLD`s, a new `mz_catalog.mz_holds` table will be maintained with structure:
+
+- `id text`: the `HOLD`'s id
+- `database_id text`
+- `schema_id text`
+- `name text`
+- `on text[]`: an array of the object ids in the `HOLD`
+- `at uint8`: the `HOLD`'s time
```

Comment:
If this happened in the same persist txn as the actual advancement, does that improve things?

> we do a time-based rate-limiting of actually letting these things go through

This sounds ok to me. It would mean that if users were SUBSCRIBE'ing to this table it wouldn't match what they were sending, and on restarts it would be a bit older than they last sent. If they were trying to aggressively get memory back it might also be detrimental there, but seems doubtful that matters much. Yeah, having some kind of limiter works here.

How does the one in persist work? LD configured? Backpressure?

Reply:
> If this happened in the same persist txn as the actual advancement, does that improve things?

Sorry, what's "the actual advancement"?

> This sounds ok to me. It would mean that if users were SUBSCRIBE'ing to this table it wouldn't match what they were sending, and on restarts it would be a bit older than they last sent. If they were trying to aggressively get memory back it might also be detrimental there, but seems doubtful that matters much. Yeah, having some kind of limiter works here.

Exactly. I hadn't thought of "they were trying to aggressively get memory back", but that's an interesting wrinkle. I _think_ the story is we'd be able to send along the advancement to clusterd regardless of the limiting, which would address memory usage. We'd just be holding back any compaction of persist shards, which doesn't really have guarantees about promptness anyway.

> How does the one in persist work? LD configured? Backpressure?

The persist one is based on an in-mem field tracking the last one it let through and the duration is derived from the read lease duration. [Code](https://github.com/MaterializeInc/materialize/blob/766d036495a4ca2427ea79d5972c57de17229068/src/persist-client/src/read.rs#L854-L873)

Reply:
> Sorry, what's "the actual advancement"?

I suppose this is two things:

1. The timestamp to which we advance the controller's read holds (which then propagate to persist/compute).
2. The timestamp that is written down for use on restart.

The builtin tables are probably not a good place to store the restart value because in other contexts those tables can always be cleared on restart. There needs to be a known safe place to recover that.

Reply:
Ah, gotcha. My understanding is that (1) is that the controller will hook these new read holds into its existing `SinceHandle::maybe_compare_and_downgrade_since` calls. I think (2) wants the same rate limiting as this `mz_catalog.mz_holds`... I actually thought mz_catalog.mz_holds _was_ (2)

Reply:
I would personally love to have linearizability with `ADVANCE HOLD` and reads from `mz_holds`. Having stale data in `mz_holds` seems pretty confusing.

Point taken for sure on the performance issues, though. Is it possible we could just slow down the `ADVANCE HOLD` command? Time only ticks forward once a second, so it doesn't seem like you'd ever be running `ADVANCE HOLD` more than once a second. Plausibly you could batch up every `ADVANCE HOLD` command that arrives in a given second and fire them all off in one batch together? Then you'd only be doing one write per second. 

Maybe 1s to run an `ADVANCE HOLD` command is too painful—but seems like there's a decent middle ground, like 3Hz or 5Hz.